examples/async/async/main.go

```diff
@@ -0,0 +1,55 @@
+package main
+
+import (
+	"context"
+	"time"
+
+	"github.com/uw-labs/substrate"
+	"github.com/uw-labs/substrate-tools/ackordering"
+	"github.com/uw-labs/substrate-tools/async"
+	"github.com/uw-labs/substrate/kafka"
+)
+
+func main() {
+	asyncSource, err := kafka.NewAsyncMessageSource(kafka.AsyncMessageSourceConfig{
+		ConsumerGroup: "example-async-sync",
+		Brokers:       []string{"localhost:9092"},
+		Topic:         "example-topic",
+		Version:       "2.0.1",
+		Offset:        -2, // oldest
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	asyncSource = ackordering.NewAsyncMessageSource(asyncSource)
+
+	opts := []async.MessageSourceOption{
+		async.WithSourceConsumers(10),     // 10 concurrent message handlers
+		async.WithSourceMsgBufferSize(10), // buffer 10 messages from the source
+		async.WithSourceAckBufferSize(10), // buffer 10 acks from the ConsumerMessageHandler
+	}
+
+	source := async.NewMessageSource(asyncSource, opts...)
+	defer func() {
+		err := source.Close()
+		if err != nil {
+			panic(err)
+		}
+	}()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	handler := func(ctx context.Context, msg substrate.Message, ack async.AckFunc) error {
+		defer ack()
+
+		println("consumed:", string(msg.Data()))
```

Comment:
```suggestion
		println("consumed:", string(msg.Data()))
                ack()
```

I think let's simplify to this, wdyt?

Reply:
also handle err from ack function, so maybe:

if err := ack(); err !=nil {
return err.
}

Reply:
As this is example code, I image most people will copy and paste it. Therefore I prefer to defer the `ack()` right at the top of the handler code - that way there's less chance of it being removed when modified resulting in broken code

Reply:
I agree with @kaperys. I appreciate an example being the bare bones to show it in a simple solution, but having it deferred is a more extendable example that people can look towards. As soon as there is branching in the handler func (if someone uses the example as a starting base), then the `ack()` will be moved. When implementing the ack func, the intention was to have it used like this in the default use cases.

Reply:
Ok I agree with defer, but then maybe let's add some error handling

Reply:
@povilasv and I had a chat offline and agreed that `return ack()` at the end of the function is probably a good approach. This way if the users message handler code returns an error and the ack() function returns an error (or, would have returned an error), the users message handler code error is returned first. What do you think @heedson?

Reply:
That's fine, but for the copy and paste use case, this will result in the same situation for the user. As soon as a 2nd `return` is added, there will be a slight bit of management added. Either is fine for an example so I'm comfortable with this.

I think it maybe is worth us/myself looking at the ack function signature and seeing if it makes sense without an error returning so this isn't even an issue in the future.

Reply:
Great! I agree that if the ack function didn't return an error orchestrating this would be simpler.